### PR TITLE
Dispose Unused Monitors

### PR DESCRIPTION
**Pull request by ColinKYuen**
### Summary

Dispose Unused Monitors

### Description

Monitors with no contexts that are being checked will be disposed from the map after a timeout set by the user in the property files. By default the timeout is 60000ms.

### Additional Reviewers

@karenc-bq 
@brunos-bq 
@sergiyv-bitquill

### CODE REVIEW

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/DefaultMonitorService.java [hunk 1]
```diff
@@ -105,9 +116,32 @@ public void releaseResources() {
...
+        value.cancel(true);
+        taskMap.remove(key);
+      }
+    });
+  }
+
   protected IMonitor getMonitor(Set<String> nodeKeys, HostInfo hostInfo, PropertySet propertySet) {
     final String node = nodeKeys.stream().filter(this.threadContainer.getMonitorMap()::containsKey).findFirst().orElse(nodeKeys.iterator().next());
     final IMonitor monitor = this.threadContainer.getMonitorMap().computeIfAbsent(node, k -> monitorInitializer.createMonitor(hostInfo, propertySet));
+    monitor.setService(this);
```
Comment (sergiyv-bitquill) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/DefaultMonitorService.java:144]: It should be an extra parameter of constructor.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/DefaultMonitorService.java [hunk 2]
```diff
@@ -47,6 +52,7 @@ public DefaultMonitorService(Log log) {
             new BasicConnectionProvider(),
             hostInfo,
             propertySet,
+            propertySet.getIntegerProperty(PropertyKey.monitorDisposeTime).getValue(),
```
Comment (brunos-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/DefaultMonitorService.java:55]: Something doesn't look right if you have to pass the whole property set, and then pass a single entry of this set again?

If the code being called get other values from the property set, can't it get this value the same way?
Or if that's the only value being fetched from the property set, why passing the whole set as parameter?

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/DefaultMonitorService.java [hunk 3]
```diff
@@ -95,7 +101,12 @@ public MonitorConnectionContext startMonitoring(
 
   @Override
   public void stopMonitoring(MonitorConnectionContext context) {
-    final IMonitor monitor = this.threadContainer.getMonitorMap().get(context.getNodeKeys().iterator().next());
+    final String node = context.getNodeKeys()
```
Comment (brunos-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/DefaultMonitorService.java:104]: NIT: add a comment somewhere in this function to say that finding any node is good/enough for this function

File: src/test/java/testsuite/failover/FailoverIntegrationTest.java [hunk 4]
```diff
@@ -73,7 +73,7 @@
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 /** Integration testing with Aurora MySQL failover logic. */
-@Disabled
+//@Disabled
```
Comment (sergiyv-bitquill) [inline src/test/java/testsuite/failover/FailoverIntegrationTest.java:76]: Why we're enabling them?

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/DefaultMonitorService.java [hunk 5]
```diff
@@ -105,9 +116,32 @@ public void releaseResources() {
...
   }
 
+  @Override
+  public synchronized void removeMonitorFromMap(IMonitor monitor) {
+    if (monitor == null) {
+      throw new NullArgumentException("Cannot remove a null monitor");
+    }
+
+    final Map<String, IMonitor> monitorMap = MonitorThreadContainer.getInstance().getMonitorMap();
+    monitorMap.forEach((key, value) -> {
```
Comment (brunos-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/DefaultMonitorService.java:126]: NIT: can you try to write it in a shorter way using a lambda?

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/IMonitor.java [hunk 6]
```diff
@@ -29,4 +29,7 @@
 public interface IMonitor extends Runnable {
   void startMonitoring(MonitorConnectionContext context);
   void stopMonitoring(MonitorConnectionContext context);
+  long getLastContextUsedTimestamp();
+  int getNumOfContexts();
```
Comment (sergiyv-bitquill) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/IMonitor.java:33]: Why do we need it?

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/DefaultMonitorService.java (hunk 1 above)
Comment (brunos-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/DefaultMonitorService.java:144]: why is this needed?

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/DefaultMonitorService.java [hunk 7]
```diff
@@ -105,9 +116,32 @@ public void releaseResources() {
     MonitorThreadContainer.releaseInstance();
   }
 
+  @Override
+  public synchronized void removeMonitorFromMap(IMonitor monitor) {
```
Comment (brunos-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/DefaultMonitorService.java:120]: Can this function be just removeMonitor() ?

Does it add something to say it's from a map?

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/IMonitorService.java [hunk 8]
```diff
@@ -43,4 +43,6 @@ MonitorConnectionContext startMonitoring(
   void stopMonitoring(MonitorConnectionContext context);
 
   void releaseResources();
+
+  void removeMonitorFromMap(IMonitor monitor);
```
Comment (brunos-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/IMonitorService.java:47]: Same comment about the function name

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/Monitor.java [hunk 9]
```diff
@@ -61,40 +62,61 @@
...
+  public synchronized void startMonitoring(MonitorConnectionContext context) {
     this.connectionCheckIntervalMillis = Math.min(
         this.connectionCheckIntervalMillis,
         context.getFailureDetectionIntervalMillis());
-
-    context.setStartMonitorTime(this.getCurrentTimeMillis());
-    contexts.add(context);
+    final long currentTime = this.getCurrentTimeMillis();
+    context.setStartMonitorTime(currentTime);
+    this.lastContextUsedTimestamp.getAndSet(currentTime);
```
Comment (brunos-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/Monitor.java:85]: NIT: don't need the "this" if your parameter name is different than the attribute.

(also, it wasn't here before? is it needed now?)

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/Monitor.java [hunk 10]
```diff
@@ -61,40 +62,61 @@
...
-  public void stopMonitoring(MonitorConnectionContext context) {
-    contexts.remove(context);
+  public synchronized void stopMonitoring(MonitorConnectionContext context) {
+    this.contexts.remove(context);
     this.connectionCheckIntervalMillis = findShortestIntervalMillis();
   }
 
+  @Override
+  public long getLastContextUsedTimestamp() {
+    return this.lastContextUsedTimestamp.get();
```
Comment (brunos-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/Monitor.java:97]: Same comment here, not sure "this" is needed

File: src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/DefaultMonitorServiceTest.java [hunk 11]
```diff
@@ -66,15 +67,15 @@
...
-  private PropertySet set;
 
+  private PropertySet set;
   private AutoCloseable closeable;
   private DefaultMonitorService monitorService;
   private ArgumentCaptor<MonitorConnectionContext> contextCaptor;
 
   @BeforeEach
   void init() {
+    set = new DefaultPropertySet();
```
Comment (brunos-bq) [inline src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/DefaultMonitorServiceTest.java:78]: set is too vague for a variable name, what about?

propertySet = new DefaultPropertySet();

File: src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/MonitorTest.java [hunk 12]
```diff
@@ -61,12 +66,23 @@
   MonitorConnectionContext contextWithLongInterval;
   @Mock
   BooleanProperty booleanProperty;
+  @Mock
+  IntegerProperty intProperty;
```
Comment (brunos-bq) [inline src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/MonitorTest.java:70]: IntegerProperty integerProperty;

File: src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/MonitorTest.java [hunk 13]
```diff
@@ -85,15 +101,47 @@ void init() throws SQLException {
     Mockito
         .when(booleanProperty.getStringValue())
         .thenReturn(Boolean.TRUE.toString());
+    Mockito
```
Comment (brunos-bq) [inline src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/MonitorTest.java:104]: static import, then remove all the "Mockito" occurences

File: src/test/java/testsuite/failover/FailoverIntegrationTest.java (hunk 4 above)
Comment (brunos-bq) [inline src/test/java/testsuite/failover/FailoverIntegrationTest.java:76]: Never great to comment stuff. Can't you just remove it?
